File: July-LeetCoding-Challenge-2020/Week2/100-Same-Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// Solution #1 DFS
// TC: O(n) number of nodes
// SC: O(h) number of times DFS called which is equal to the height of the tree 
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 10.1 MB, less than 15.43%
class Solution {
    bool same = true;
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        
        DFS(p,q);
        
        return same;
    }
    
    void DFS(TreeNode *p, TreeNode *q){
        if(!p || !q){
            if(!(!p && !q))
                same = false;
            return;
        }
        else if(p->val != q->val)
            same = false;
        
        isSameTree(p->left, q->left);
        isSameTree(p->right, q->right);
    }
};

File: July-LeetCoding-Challenge-2020/Week2/100-same-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # Sol 1. Recursion
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        if not p and not q:
            return True
    
        if not p or not q:
            return False

        if p.val != q.val:
            return False
        
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
    # TC: O(N), minimum number of nodes in p or q
    # SC: O(h), where h is the shorter depth in p or q, h represents the height of a tree
    # Runtime: 28 ms, faster than 81.78%
    # Memory Usage: 13.9 MB, less than 24.53%
